File: lib/core_services/eiga/ab_eiga_service.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:hoyomi/core_services/eiga/main.dart';

export 'package:hoyomi/core_services/service.dart';

part 'ab_eiga_service.freezed.dart';
part 'ab_eiga_service.g.dart';

abstract class ABEigaService extends Service
    with EigaFollowGeneralMixin, EigaWatchTimeGeneralMixin {
  Future<SourceContent> fetchSourceContent({required SourceVideo source}) {
    throw UnimplementedError();
  }

  Future<EigaCategory> getCategory({
    required String categoryId,
    required int page,
    required Map<String, List<String>?> filters,
  });

  Future<EigaCategory> getExplorer({
    required int page,
    required Map<String, List<String>?> filters,
  }) {
    throw UnimplementedError();
  }

  Future<MetaEiga> getDetails(String eigaId);

  Future<EigaEpisodes> getEpisodes(String eigaId);

  Future<OpeningEnding?> getOpeningEnding(EigaSourceContext context) {
    throw UnimplementedError();
  }

  Future<Vtt?> getSeekThumbnail(EigaSourceContext context) {
    throw UnimplementedError();
  }

  Future<List<ServerSource>> getServers({
    required String eigaId,
    required EigaEpisode episode,
  }) {
    throw UnimplementedError();
  }

  Future<SourceVideo> getSource({
    required String eigaId,
    required EigaEpisode episode,
    ServerSource? server,
  });

  Future<List<Subtitle>> getSubtitles({
    required String eigaId,
    required EigaEpisode episode,
    required SourceVideo source,
  }) {
    throw UnimplementedError();
  }

  Future<List<Eiga>> getSuggest({
    required MetaEiga metaEiga,
    required String eigaId,
    int? page,
  }) {
    throw UnimplementedError();
  }

  Future<String> getURL(String eigaId, {String? episodeId}) {
    throw UnimplementedError();
  }

  Future<EigaHome> home();

  Future<EigaCategory> search({
    required String keyword,
    required int page,
    required Map<String, List<String>?> filters,
    required bool quick,
  });
}

@freezed
sealed class EigaSourceContext with _$EigaSourceContext {
  const factory EigaSourceContext({
    required String eigaId,
    required MetaEiga metaEiga,
    required EigaEpisode episode,
    required SourceVideo source,
  }) = _EigaSourceContext;

  factory EigaSourceContext.fromJson(Map<String, dynamic> json) =>
      _$EigaSourceContextFromJson(json);
}

File: lib/core_services/eiga/ab_eiga_service.ts
import { BaseService } from "../service"
import {
  type MetaEiga,
  type EigaEpisode,
  type SourceVideo,
  type EigaHome,
  type EigaCategory,
  type EigaEpisodes,
  type ServerSource,
  type SourceContent,
  type Subtitle,
  type Vtt,
  type OpeningEnding,
  type Eiga
} from "../main"

export interface EigaSourceContext {
  eigaId: string
  metaEiga: MetaEiga
  episode: EigaEpisode
  source: SourceVideo
}

// Abstract class ABEigaService
export abstract class ABEigaService extends BaseService {
  readonly type = "eiga"

  abstract getURL(eigaId: string, chapterId?: string): Promise<string>

  abstract home(): Promise<EigaHome>

  abstract getCategory(params: {
    categoryId: string
    page: number
    filters: { [key: string]: string[] | null }
  }): Promise<EigaCategory>

  getExplorer(
    page: number,
    filters?: { [key: string]: string[] | null }
  ): Promise<EigaCategory> {
    throw new UnimplementedError()
  }

  abstract getDetails(eigaId: string): Promise<MetaEiga>
  abstract getEpisodes(eigaId: string): Promise<EigaEpisodes>

  getServers(params: {
    eigaId: string
    episode: EigaEpisode
  }): Promise<ServerSource[]> {
    throw new UnimplementedError()
  }

  abstract getSource(params: {
    eigaId: string
    episode: EigaEpisode
    server?: ServerSource
  }): Promise<SourceVideo>

  fetchSourceContent(source: SourceVideo): Promise<SourceContent> {
    throw new UnimplementedError()
  }

  getSubtitles(params: {
    eigaId: string
    episode: EigaEpisode
    source: SourceVideo
  }): Promise<Subtitle[]> {
    throw new UnimplementedError()
  }

  getSeekThumbnail(context: EigaSourceContext): Promise<Vtt | null> {
    throw new UnimplementedError()
  }

  getOpeningEnding(context: EigaSourceContext): Promise<OpeningEnding | null> {
    throw new UnimplementedError()
  }

  getSuggest(params: {
    metaEiga: MetaEiga
    eigaId: string
    page?: number
  }): Promise<Eiga[]> {
    throw new UnimplementedError()
  }

  abstract search(params: {
    keyword: string
    page: number
    filters: { [key: string]: string[] | null }
    quick: boolean
  }): Promise<EigaCategory>
}
